src/utilities/bmi/mass_balance.cpp

```diff
@@ -0,0 +1,137 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+------------------------------------------------------------------------
+
+Version 0.1
+Implementation the BMI mass balance checking protocol
+*/
+
+#include "mass_balance.hpp"
+
+namespace models { namespace bmi { namespace protocols {
+
+void NgenMassBalance::run(const Context& ctx) const{
+    bool check_step = false;
+    //if frequency was set to -1, only check at the end
+    if( frequency == -1 ){
+        if(ctx.current_time_step == ctx.total_steps){
+            check_step = true;
+        }
+    }
+    else{
+        check_step = (ctx.current_time_step % frequency) == 0;
+    }
+
+    if(model && is_supported && check && check_step) {
+        double mass_in, mass_out, mass_stored, mass_leaked, mass_balance;
+        model->GetValue(INPUT_MASS_NAME, &mass_in);
+        model->GetValue(OUTPUT_MASS_NAME, &mass_out);
+        model->GetValue(STORED_MASS_NAME, &mass_stored);
+        model->GetValue(LEAKED_MASS_NAME, &mass_leaked);
+        // TODO consider unit conversion if/when it becomes necessary
+        mass_balance = mass_in - mass_out - mass_stored - mass_leaked;
+        if ( std::abs(mass_balance) > tolerance ) {
+            std::stringstream ss;
+            ss << "at timestep " << std::to_string(ctx.current_time_step)
+               << " ("+ctx.timestamp+")"
+               << " at feature id " << ctx.id <<std::endl
+               << "\tMass balance check failed for " << model->GetComponentName() << "\n\t" <<
+                INPUT_MASS_NAME << "(" << mass_in <<  ") - " <<
+                OUTPUT_MASS_NAME << " (" << mass_out <<  ") - " << 
+                STORED_MASS_NAME << " (" << mass_stored <<  ") - " << 
+                LEAKED_MASS_NAME << " (" << mass_leaked << ") = " << 
+                mass_balance << "\n\t" << "tolerance: " << tolerance << std::endl;
+            if(is_fatal)
+                throw MassBalanceError("mass_balance::error " + ss.str());
+            else{
+                std::cerr << "mass_balance::warning " + ss.str();
+            }
+                
+        }
+    }
+}
+
+void NgenMassBalance::check_support() {
+    if (model->is_model_initialized()) {
+        double mass_var;
+        std::vector<std::string> units;
+        units.reserve(4);
+        try{
+            for(const auto& name : 
+                {INPUT_MASS_NAME, OUTPUT_MASS_NAME, STORED_MASS_NAME, LEAKED_MASS_NAME}
+            ){
+                model->GetValue(name, &mass_var);
+                units.push_back( model->GetVarUnits(name) );
+            }
+            //Compare all other units to the first one (+1)
+            if( std::equal( units.begin()+1, units.end(), units.begin() ) ) {
```

Comment:
Unless I've missed something (entirely possible; if so, sorry), there are no explicit units defined anywhere for `tolerance`, and they are simply taken to be whatever units that particular model is using.  This seems like an easy place for the wrong assumption to be made.  We might want to make tolerance units an explicit part of the config and protocol, and then check those here as well (though perhaps slightly separately so a distinctive error message can be used for that).

Not unrelated, a small configuration example and some discussion within _doc/BMI_MODELS.md_ would probably be helpful to include as part of the overall changes.